proto_parser/demo/demo_main.cc

```diff
@@ -0,0 +1,395 @@
+  //                 # Condition 3.2
+  //                 index_ui {
+  //                   index: -2
+  //                   path: "type"
+  //                   equals_int: 1  # SOMETHING
+  //                 }
+  //               }
+  //             }
+  //           }
+  std::unique_ptr<ProtoValue> ConvertEventPredicateFromFile(
+      absl::string_view path) {
+    ProtoParser parser(ReadTextProtoFromStream(path));
+    EventPredicate event_predicate_prototype;
+    std::unique_ptr<ProtoValue> and_field = 
+        absl::make_unique<MessageValue>("and", 0, 0);
+    MessageValue* and_field_ptr = static_cast<MessageValue*>(and_field.get());
+
+    std::unique_ptr<ProtoValue> event_predicate = 
+        parser.Create(event_predicate_prototype);
+    MessageValue* event_predicate_ptr = 
+        static_cast<MessageValue*>(event_predicate.get());
+    const std::vector<std::unique_ptr<ProtoValue>>& fields = 
+      event_predicate_ptr->GetFields();
+
+    for (const std::unique_ptr<ProtoValue>& field : fields) {
+      if (field.get()->GetName() == "event_matcher") {
+        absl::flat_hash_map<std::string, ProtoValue*> name_to_ptr_map = 
+            GetFieldNameToPtrMap(*field.get());
+        if (name_to_ptr_map.find("int_comparator") == name_to_ptr_map.end()) {
+          and_field_ptr->AddField(EventMatcherToFieldExists(*field.get()));
+        } else {
+          and_field_ptr->AddField(EventMatcherToFieldValue(*field.get()));
+        }
+        
+      } else if (field.get()->GetName() == "ui_element_path_predicate") {
+        and_field_ptr->AddField(UiElementPathPredicateToUiElement(*field.get()));
+      }
+    }
+
+    return and_field;
+  }
+
+ private:
+
+  // Reads text proto into a string from a file.
+  std::string ReadTextProtoFromStream(absl::string_view path) {
+    std::ifstream ifs;
+    std::string text_proto;
+
+    ifs.open(std::string(path));
+
+    assert(ifs.good());
+
+    while (ifs.good()) {
+      std::string tmp;
+      std::getline(ifs, tmp);
+      text_proto += tmp + "\n";
+    }
+
+    return text_proto;
+  }
+
+  // Copies the comments of an original proto value to a copy, if they're both
+  // primitive types also copy the value.
+  void CopyVal(ProtoValue* copy, const ProtoValue& original) {
+    copy->SetCommentBehindField(original.GetCommentBehindField());
+    copy->SetCommentAboveField(original.GetCommentAboveField());
+    PrimitiveValue* copy_val = dynamic_cast<PrimitiveValue*>(copy);
+    const PrimitiveValue* original_val = dynamic_cast<const PrimitiveValue*>(
+        &original);
+    if (copy_val != nullptr && original_val != nullptr) {
+      // Copy their values only if they are both primitive values.
+      copy_val->SetVal(original_val->GetVal());
+    }
+  }
+
+  // Gets a hash table that maps field names of a message value to their
+  // corresponding protovalue pointer.
+  absl::flat_hash_map<std::string, ProtoValue*> GetFieldNameToPtrMap(
+      const ProtoValue& original) {
```

Comment:
const MessageValue& original

since you will convert it into MessageValue anyway.